src/OptimalSwapRouter.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.18;
+
+import "solady/src/utils/SafeTransferLib.sol";
+import {INonfungiblePositionManager} from "@aperture_finance/uni-v3-lib/src/interfaces/INonfungiblePositionManager.sol";
+import {PoolAddress, PoolKey} from "@aperture_finance/uni-v3-lib/src/PoolAddress.sol";
+import {TernaryLib} from "@aperture_finance/uni-v3-lib/src/TernaryLib.sol";
+import {ERC20Callee} from "./libraries/ERC20Caller.sol";
+import {OptimalSwap, V3PoolCallee} from "./libraries/OptimalSwap.sol";
+import {Payments, SwapRouter, UniV3Immutables} from "./base/SwapRouter.sol";
+
+/// @title Optimal Swap Router
+/// @author Aperture Finance
+/// @dev This router swaps through an aggregator to get to approximately the optimal ratio to add liquidity in a Uni V3
+/// pool, then swaps the tokens to the optimal ratio to add liquidity in the same pool.
+contract OptimalSwapRouter is UniV3Immutables, Payments, SwapRouter {
+    using SafeTransferLib for address;
+    using TernaryLib for bool;
+
+    constructor(INonfungiblePositionManager npm) UniV3Immutables(npm) {}
+
+    fallback() external {
```

Comment:
Since we use compact encoding and decode manually, we don't need a function interface and resort to Solidity's codegen for ABI decoding. Also we shouldn't need more than one entry point to the router.

As far as I conceive, it's permissionless and stateless. Need to think about if DoS is possible though.